### PR TITLE
Add KafkaUI. Add a kafka playground too.

### PR DESCRIPTION
**Pull request by g7ed6e**
Add a Kafka playground.
Add support of KafkaUI to kafka resource builder.

Usage
```c#
var kafka = builder.AddKafka("kafka")
    .WithKafkaUi(8080);
```

Implementation notes
A Kafka broker node needs to know which name is used by clients to connect to itself to properly work. Please see https://www.confluent.io/blog/kafka-listeners-explained/. This subtlety leads to expose 2 endpoints in Aspire. One for host processes and one for containers.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3882)

### CODE REVIEW

File: src/Aspire.Hosting.Kafka/KafkaContainerImageTags.cs [hunk 1]
```diff
@@ -8,4 +8,6 @@ internal static class KafkaContainerImageTags
     public const string Registry = "docker.io";
     public const string Image = "confluentinc/confluent-local";
     public const string Tag = "7.6.0";
+    public const string KafkaUiImage = "provectuslabs/kafka-ui";
```
Comment (davidfowl) [inline src/Aspire.Hosting.Kafka/KafkaContainerImageTags.cs:11]: Is this the defacto standard?

**Comment by davidfowl** [comment]
@g7ed6e can you update the description?

File: src/Aspire.Hosting.Kafka/KafkaContainerImageTags.cs (hunk 1 above)
Comment (g7ed6e) [reply src/Aspire.Hosting.Kafka/KafkaContainerImageTags.cs:11]: This the UI suggested by this tutorial https://learn.microsoft.com/en-us/azure/container-apps/tutorial-dev-services-kafka?tabs=bash

**Comment by g7ed6e** [comment]
> @g7ed6e can you update the description?

@davidfowl done. I added a WithSchemaRegistry extension method.

File: src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs [hunk 2]
```diff
@@ -25,11 +28,68 @@ public static IResourceBuilder<KafkaServerResource> AddKafka(this IDistributedAp
...
+    /// <summary>
+    /// Adds a Schema Registry container to the application. This version of the package defaults to the 7.6.0 tag of the confluentinc/cp-schema-registry container image.
+    /// </summary>
+    /// <param name="builder"></param>
+    /// <param name="hostPort"></param>
+    /// <param name="containerName"></param>
+    /// <returns></returns>
+    public static IResourceBuilder<KafkaServerResource> WithSchemaRegistry(this IResourceBuilder<KafkaServerResource> builder, int? hostPort = null, string? containerName = null)
+    {
+        if (builder.ApplicationBuilder.Resources.OfType<SchemaRegistryContainerResource>().Any())
```
Comment (davidfowl) [inline src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs:73]: Why can there only be one?

File: src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs [hunk 3]
```diff
@@ -25,11 +28,68 @@ public static IResourceBuilder<KafkaServerResource> AddKafka(this IDistributedAp
...
+    }
+
+    /// <summary>
+    /// Adds a Schema Registry container to the application. This version of the package defaults to the 7.6.0 tag of the confluentinc/cp-schema-registry container image.
+    /// </summary>
+    /// <param name="builder"></param>
+    /// <param name="hostPort"></param>
+    /// <param name="containerName"></param>
+    /// <returns></returns>
+    public static IResourceBuilder<KafkaServerResource> WithSchemaRegistry(this IResourceBuilder<KafkaServerResource> builder, int? hostPort = null, string? containerName = null)
```
Comment (davidfowl) [inline src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs:71]: We don’t use hostPort anymore, it’s port

File: src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs [hunk 4]
```diff
@@ -25,11 +28,68 @@ public static IResourceBuilder<KafkaServerResource> AddKafka(this IDistributedAp
...
+    {
+        if (builder.ApplicationBuilder.Resources.OfType<SchemaRegistryContainerResource>().Any())
+        {
+            return builder;
+        }
+
+        builder.ApplicationBuilder.Services.TryAddLifecycleHook<SchemaRegistryConfigWriterHook>();
```
[the remark this answers is not in the document]
Comment (g7ed6e) [reply src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs]: There's is no longer a hook.

File: src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs (hunk 2 above)
Comment (g7ed6e) [reply src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs]: That's right there may be several.

**Comment by g7ed6e** [comment]
SchemaRegistry related updates have been removed from that PR as I think it requires more experimentations. (client side code need to feed the `ProducerBuilder<TKey, TValue>` / `Consumer<TKey, TValue>` with schema registry url.. please see https://github.com/confluentinc/confluent-kafka-dotnet/blob/c251c6c486cc4b6283deda87be4cd9de9d2ffd12/examples/JsonSerialization/Program.cs#L139).

File: src/Aspire.Hosting.Kafka/KafkaServerResource.cs [hunk 5]
```diff
@@ -12,14 +12,21 @@ namespace Aspire.Hosting;
 public class KafkaServerResource(string name) : ContainerResource(name), IResourceWithConnectionString, IResourceWithEnvironment
 {
     internal const string PrimaryEndpointName = "tcp";
+    internal const string InternalEndpointName = "internal";
```
Comment (davidfowl) [inline src/Aspire.Hosting.Kafka/KafkaServerResource.cs:15]: Add a comment explaining what each of these endpoints represent.

File: src/Aspire.Hosting.Kafka/SchemaRegistryContainerResource.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+
+namespace Aspire.Hosting;
+
+internal sealed class SchemaRegistryContainerResource(string name) : ContainerResource(name)
```
Comment (davidfowl) [inline src/Aspire.Hosting.Kafka/SchemaRegistryContainerResource.cs:8]: You can remove this now.

File: src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs [hunk 7]
```diff
@@ -25,11 +26,34 @@ public static IResourceBuilder<KafkaServerResource> AddKafka(this IDistributedAp
         var kafka = new KafkaServerResource(name);
         return builder.AddResource(kafka)
             .WithEndpoint(targetPort: KafkaBrokerPort, port: port, name: KafkaServerResource.PrimaryEndpointName)
+            .WithEndpoint(targetPort: 9093, name: KafkaServerResource.InternalEndpointName)
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs:29]: Work having a constant for this too? Makes it more self documenting?

Comment (g7ed6e) [reply src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs]: added a constant an used it.

File: src/Aspire.Hosting.Kafka/SchemaRegistryContainerResource.cs (hunk 6 above)
Comment (g7ed6e) [reply src/Aspire.Hosting.Kafka/SchemaRegistryContainerResource.cs]: good catch. thanks!

File: src/Aspire.Hosting.Kafka/KafkaServerResource.cs (hunk 5 above)
Comment (g7ed6e) [reply src/Aspire.Hosting.Kafka/KafkaServerResource.cs:17]: done. Extended public xml doc too.

**Comment by davidfowl** [comment]
LGTM

File: src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs [hunk 8]
```diff
@@ -25,11 +27,34 @@ public static IResourceBuilder<KafkaServerResource> AddKafka(this IDistributedAp
...
     }
 
+    /// <summary>
+    /// Adds a Kafka UI container to the application. This version of the package defaults to the 0.7.2 tag of the provectuslabs/kafka-ui container image.
+    /// </summary>
+    /// <param name="builder"></param>
+    /// <param name="port"></param>
+    /// <param name="containerName"></param>
+    /// <returns></returns>
+    public static IResourceBuilder<KafkaServerResource> WithKafkaUI(this IResourceBuilder<KafkaServerResource> builder, int? port = null, string? containerName = null)
```
Comment (davidfowl) [inline src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs:43]: Ah I just realized this is missing the overload that exposes the `KafkaUIContainerResource` that allows others to change container properties (tag etc). https://github.com/dotnet/aspire/blob/f552a0d94409f0e605a253a38f93da3a5be6b215/src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs#L97

Comment (g7ed6e) [reply src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs]: Updated.

@davidfowl `WithPgAdmin` have a nice capability which allows to use a single instance to manage multiple resources. Should we implement something similar to manage several Kafka resources in a unique KafkaUI container ?
`KafkaBuilderExtensions` and related resources live in `Aspire.Hosting` namespace. I noticed this is not consistent accross components (see MongoDB or Postgres resources for example). What is the correct namespace rule ?

Comment (davidfowl) [reply src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs]: Yes we should. Extensions should all live in the Aspire.Hosting namespace.

cc @mitchdenny

File: playground/kafka/Consumer/Program.cs [hunk 9]
```diff
@@ -6,8 +6,10 @@
 
 var builder = Host.CreateApplicationBuilder(args);
 
+var kafka = args[0];
```
Comment (davidfowl) [inline playground/kafka/Consumer/Program.cs:9]: none of our samples do this. I would revert to make it consistent with what they are doing.

Comment (g7ed6e) [reply playground/kafka/Consumer/Program.cs:9]: I will do. Is this supported or is it a side effect ?
It brings ability to write the following apphost which I find interesting

```c#
var builder = DistributedApplication.CreateBuilder(args);

var kafka = builder.AddKafka("kafka")
    .WithKafkaUI(kafkaUi => kafkaUi.WithHostPort(8080));

builder.AddProject<Projects.Producer>("producer")
    .WithReference(kafka)
    .WithArgs(kafka.Resource.Name);

builder.AddProject<Projects.Consumer>("consumer")
    .WithReference(kafka)
    .WithArgs(kafka.Resource.Name);

var kafka2 = builder.AddKafka("kafka2").WithKafkaUI();

builder.AddProject<Projects.Producer>("producer2")
    .WithReference(kafka2)
    .WithArgs(kafka2.Resource.Name);

builder.AddProject<Projects.Consumer>("consumer2")
    .WithReference(kafka2)
    .WithArgs(kafka2.Resource.Name);
```

Comment (g7ed6e) [reply playground/kafka/Consumer/Program.cs]: @davidfowl reverted

File: src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs (hunk 8 above)
Comment (g7ed6e) [reply src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs]: done

File: src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs [hunk 10]
```diff
@@ -25,11 +29,60 @@ public static IResourceBuilder<KafkaServerResource> AddKafka(this IDistributedAp
...
+    /// <param name="builder">The Kafka server resource builder.</param>
+    /// <param name="configureContainer">Configuration callback for KafkaUI container resource.</param>
+    /// <param name="containerName">The name of the container (Optional).</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{KafkaServerResource}"/>.</returns>
+    public static IResourceBuilder<KafkaServerResource> WithKafkaUI(this IResourceBuilder<KafkaServerResource> builder, Action<IResourceBuilder<KafkaUIContainerResource>>? configureContainer = null, string? containerName = null)
+    {
+        if (builder.ApplicationBuilder.Resources.OfType<KafkaUIContainerResource>().SingleOrDefault() is { } existingKafkaUIResource)
+        {
+            var builderForExistingResource = builder.ApplicationBuilder.CreateResourceBuilder(existingKafkaUIResource);
+            configureContainer?.Invoke(builderForExistingResource);
```
[the remark this answers is not in the document]
Comment (g7ed6e) [reply src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs:50]: With [Postgres ](https://github.com/dotnet/aspire/blob/b5b22fe36c44667aa50aaee4e1fc27a3d67b5729/src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs#L76)and [Redis ](https://github.com/dotnet/aspire/blob/b5b22fe36c44667aa50aaee4e1fc27a3d67b5729/src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs#L76 )we do. 
With [MySQL ](https://github.com/dotnet/aspire/blob/b5b22fe36c44667aa50aaee4e1fc27a3d67b5729/src/Aspire.Hosting.MySql/MySqlBuilderExtensions.cs#L67)we don't. 
With [Mongo ](https://github.com/dotnet/aspire/blob/b5b22fe36c44667aa50aaee4e1fc27a3d67b5729/src/Aspire.Hosting.MongoDB/MongoDBBuilderExtensions.cs#L62)we start one MongoExpress container instance per MongoDB resource. 
With [RabbitMQ ](https://github.com/dotnet/aspire/blob/b5b22fe36c44667aa50aaee4e1fc27a3d67b5729/src/Aspire.Hosting.RabbitMQ/RabbitMQBuilderExtensions.cs#L88)we modify the user provided image use one that embed a management plugin.
With Oracle and MS SQL Server we don't have a UI / management feature.

**Review by davidfowl — APPROVED** [review]

File: src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (mitchdenny) [reply src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs:50]: We should just create one resource but happy for us to take it as a follow up in the interests of getting this in. I've created on for PhpMyAdmin.

**Review by mitchdenny — APPROVED** [review]

**Comment by slang25** [comment]
This looks great. What are the timelines on the Aspire 8.1 release out of interest?

**Comment by slang25** [comment]
Thanks, I was able to easily retrofit this PR onto the current version so am happy to wait ☺️

**Pull request merged**